### DISCUSSION TITLE
Testing database interactions

### DISCUSSION THREAD

**Discussion by Inayat-C** [prisma/prisma#2869] [Q&A] [answered]
Hi all,

I'm fairly new to testing and have been playing around with `jest` and `supertest` to test my endpoints. Doing so brought up the question if wether i should also be testing my database interactions?

Consider the following example:

We have a generic /register endpoint in express. I'm running a series of test which should meet the following conditions: 

1) If a request with valid data is sent to this endpoint, then it should return a http 200 code with a JSON object containing the user email and user ID. Pretty generic stuff.

2) If a request with invalid data is sent (e.g fails during my validation middleware), it should return a http 400 with the error message.

These 2 tests would give me enough confidence that my endpoint will function as intended, but what if at some point my database interaction failed? So for example if we consider the following:

```js
module.exports = async (req, res) => {
	try {
		// Checks to see if a user with this email address already exists.
		const user = await prisma.users.findOne({
			where: {
				userEmail: req.body.email,
			},
		});

		/**
		 * Creates a new user record in the database if the user
		 * doesn't exist, otherwise throws an error.
		 */
		if (!user) {
			const newUser = await prisma.users.create({
				data: {
					userEmail: req.body.email,
					userPassword: passwordHash,
				},
			});

			res.status(200).json({
				email: newUser.userEmail,
				userID: newUser.id,
			});
		} else {
			throw new Error(
				"An account using this email address has already been registered."
			);
		}
	} catch (e) {
		res.status(400).json({ msg: e.message });
	}
};
```

We know that according to the docs `findOne` returns an `object` or `null`. So in our case, if a user already exists in our database, our function would return `null` and would move on to our `catch` block which would return a valid error. 

Now my question is, would there be a need to test this database interaction? Similarly with the `create` method, if for any reason it fails to create a new user record, it should throw a valid error? 

Maybe my example is too basic and there might actually be more complex use cases where testing Prisma database interactions would be desirable, but interested to hear thoughts on this and how one might go about doing so!

**Comment by ryands17** [comment, marked as answer]
Hey @Inayat-C :wave: 
I have created [tests](https://github.com/ryands17/nexus-framework-auth) with Nexus in this repo, but you can use supertest with Express in the same manner.

In [this](https://github.com/ryands17/nexus-framework-auth/blob/master/tests/user.test.ts) file, I have checked for a user signing up with the same details again, so you could add an expect in the same manner.

**Reply by Inayat-C** [reply]
Hey @ryands17 
Sorry if i get this wrong, but that's just testing your endpoints though isn't it? My query was more towards testing the database interaction with the Prisma Client. 

Should we be adding tests for when the Prisma client performs CRUD operations on the database?

**Reply by ryands17** [reply]
You can do that as well by just exporting the Prisma client and testing the values returned are equar or present. The example I have provided is more of an integration test and that in turn also tests the Prisma interactions with the database as well (because that would be what you would be writing in the resolvers anyway).

**Reply by Inayat-C** [reply]
Gotcha, thanks ;). Seeing your example helped clear some things up!

**Comment by ludralph** [comment]
Hi @Inayat-C,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
